apps/finance/contracts/Finance.sol

```diff
@@ -166,32 +160,35 @@ contract Finance is AragonApp, ERC677Receiver {
         external
     {
         require(msg.sender == _token);
-        ERC20 token = ERC20(msg.sender);
         _recordIncomingTransaction(
-            token,
+            _token,
             _from,
             _amount,
             string(_data)
         );
-        require(token.transferFrom(_from, address(vault), _amount));
+        // first we need to get the tokens to Finance
+        ERC20(_token).transferFrom(_from, this, _amount);
+        // and then approve them to vault
+        ERC20(_token).approve(address(vault), _amount);
+        // finally we can deposit them
+        vault.deposit(_token, this, _amount, new bytes(0));
     }
 
     /**
-    * @dev Deposit for ERC677 tokens
-    * @param from Address sending the tokens
-    * @param amount Amount of tokens sent
-    * @param data Data payload being executed (payment reference)
+    * @dev Deposit for ERC777 tokens
+    * @param _from Address sending the tokens
+    * @param _amount Amount of tokens sent
+    * @param _userData Data payload being executed (payment reference)
+    * @param _operatorData Data payload
     */
-    function tokenFallback(address from, uint256 amount, bytes data) transitionsPeriod external returns (bool success) {
-        ERC20 token = ERC20(msg.sender);
+    function tokenReceived(address _from, uint256 _amount, bytes _userData, bytes _operatorData) transitionsPeriod external {
         _recordIncomingTransaction(
-            token,
-            from,
-            amount,
-            string(data)
+            msg.sender,
+            _from,
+            _amount,
+            string(_userData)
         );
-        require(token.transfer(address(vault), amount));
-        return true;
+        vault.deposit(msg.sender, _from, _amount, _userData);
```

Comment:
Doesn't finance need to create an allowance or setOperator here for this to work?

Reply:
Yes, well, as ERC777 is not part of the release and Vault doesn't fully support it yet, there was no tests for it .. so even the name and the signature are wrong! :(
Wouldn't something like this work?
```
ERC777(msg.sender).send(adress(vault), _amount, _userData);
```
Anyway, maybe I could just comment it out until we have `ERC777` on Vault.